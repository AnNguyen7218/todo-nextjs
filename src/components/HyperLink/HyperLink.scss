@import 'config.colors';

.link {
  text-decoration: none;
  display: inline-block;

  &--white {
    $link-color: $white;
    color: $link-color;
    &:active {
      color: $link-color;
    }
  }

  &--purple {
    $link-color: $purple;
    color: $link-color;
    &:active {
      color: $link-color;
    }
  }

  &--hover {
    opacity: 0.5;
    &:hover {
      opacity: 1;
    }
  }

  &--underline {
    text-decoration: underline;
  }
}
