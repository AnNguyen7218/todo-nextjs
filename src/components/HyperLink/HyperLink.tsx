import React from 'react';
import cx from 'classnames';
import Link from 'next/link';
import css from './HyperLink.scss';

type Props = React.DetailedHTMLProps<
  React.AnchorHTMLAttributes<HTMLAnchorElement>,
  HTMLAnchorElement
> & {
  href: string;
  children: React.ReactNode;
  className?: string;
  theme?: 'white' | 'purple';
  hasHoverOpacity?: boolean;
  isUnderline?: boolean;
};

const HyperLink: React.FC<Props> = ({
  href,
  children,
  className,
  theme = 'white',
  hasHoverOpacity = true,
  isUnderline = false,
  ...props
}) => {
  let isExternal = false;
  if (/^http/.test(href) || /^mailto/.test(href)) {
    isExternal = true;
  }

  const linkComponentProps = isExternal ? { href, ...props } : props;
  const linkComponent = () => (
    <a
      className={cx(
        css['link'],
        css[`link--${theme}`],
        {
          [css['link--hover']]: hasHoverOpacity,
          [css['link--underline']]: isUnderline,
        },
        className
      )}
      {...linkComponentProps}
    >
      {children}
    </a>
  );

  if (isExternal) {
    return linkComponent();
  }

  return <Link href={href}>{linkComponent()}</Link>;
};

export default HyperLink;
