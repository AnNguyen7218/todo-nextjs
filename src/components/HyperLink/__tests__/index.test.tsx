import React from 'react';
import { render, cleanup } from '@testing-library/react';
import HyperLink from '../HyperLink';

afterEach(cleanup);

describe('HyperLink Component', () => {
  test('HyperLink should match snapshot', () => {
    const { container } = render(
      <>
        <HyperLink href="https://google.com">External Link</HyperLink>
        <HyperLink href="/about">Internal Link</HyperLink>
        <HyperLink href="/about" theme="purple">
          Internal Link
        </HyperLink>

        <HyperLink href="/about" theme="purple" hasHoverOpacity={false}>
          Internal Link
        </HyperLink>

        <HyperLink
          href="/about"
          theme="purple"
          hasHoverOpacity={false}
          isUnderline
        >
          Internal Link
        </HyperLink>

        <HyperLink href="/about" theme="purple" isUnderline>
          Internal Link
        </HyperLink>

        <HyperLink href="/about" theme="white" isUnderline>
          Internal Link
        </HyperLink>
      </>
    );
    expect(container).toMatchSnapshot();
  });
});
