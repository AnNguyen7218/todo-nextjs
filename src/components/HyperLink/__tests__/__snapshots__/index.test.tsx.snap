// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HyperLink Component HyperLink should match snapshot 1`] = `
<div>
  <a
    class="link link--white link--hover"
    href="https://google.com"
  >
    External Link
  </a>
  <a
    class="link link--white link--hover"
    href="/about"
  >
    Internal Link
  </a>
  <a
    class="link link--purple link--hover"
    href="/about"
  >
    Internal Link
  </a>
  <a
    class="link link--purple"
    href="/about"
  >
    Internal Link
  </a>
  <a
    class="link link--purple link--underline"
    href="/about"
  >
    Internal Link
  </a>
  <a
    class="link link--purple link--hover link--underline"
    href="/about"
  >
    Internal Link
  </a>
  <a
    class="link link--white link--hover link--underline"
    href="/about"
  >
    Internal Link
  </a>
</div>
`;
